### PR TITLE
KIWI-1256-QA Added Authorisation tests

### PR DESCRIPTION
**Pull request by shenry99**
The creation of Authoristaion Happy and Unhappy Path tests for BAV API

Added Happy and Unhappy path tests to test the /authorisation end point

### Issue tracking

- [KIWI-1256](https://govukverify.atlassian.net/browse/KIWI-1256)

### PII logging
- [ ] Verified that no PII data is being logged

### Environment variables or secrets
- [ ] No environment variables or secrets were added or changed

### Other considerations
N/A

[KIWI-1256]: https://govukverify.atlassian.net/browse/KIWI-1256?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/tests/api/BavHappyPath.test.ts [hunk 1]
```diff
@@ -41,8 +47,27 @@ describe("Test BAV End Points", () => {
...
+        const authCode = authResponse.data.authorizationCode;
+        const authRepeatResponse = await authorizationGet(origSessionId);
+        const authRepeatResponseCode = authRepeatResponse.data.authorizationCode;
+        expect(authCode).not.toEqual(authRepeatResponseCode); 
+    });
+});
+
 describe("E2E Happy Path Well Known Endpoint", () => {
-    it("E2E Happy Path Journey - Well Known", async () => {
+    it.skip("E2E Happy Path Journey - Well Known", async () => {
```
Comment (Caitlin-cooling) [inline src/tests/api/BavHappyPath.test.ts:70]: Why do we need to skip this? I thought the well known endpoint was finished

File: src/tests/api/BavUnhappyPath.test.ts [hunk 2]
```diff
@@ -10,17 +12,32 @@ describe("/session Unhappy Path", () => {
...
+
+});
+
+describe("E2E Unhappy Path /authorisation Endpoint", () => {
+	let sessionId: any;
+	beforeEach(async () => {
+		sessionId = await startStubServiceAndReturnSessionId(bavStubPayload);
+	});
+
+	it.skip("Incorrect session state", async () => {
```
Comment (Caitlin-cooling) [inline src/tests/api/BavUnhappyPath.test.ts:38]: And this? I would have hoped it would work

File: src/tests/api/BavHappyPath.test.ts [hunk 3]
```diff
@@ -16,22 +17,27 @@ describe("Test BAV End Points", () => {
 
     beforeEach(async () => {
         //Session Request
-        sessionId = await startStubServiceAndReturnSessionId(bavStubPayload);
+       sessionId = await startStubServiceAndReturnSessionId(bavStubPayload);
     });
-    it("E2E BAV End Points Happy Path Journey", async () => {
+    it.skip("E2E BAV End Points Happy Path Journey", async () => {
```
Comment (Z-Mohamoud) [inline src/tests/api/BavHappyPath.test.ts:22]: Why has the /session validation been removed? 

We should not be removing any existing logic and adding to it instead. 

Please re-add the authSessionState validation that was part of the /session endpoint PR and then you should kick off your logic to hit the /authorization endpoint and then validate the authSessionState again

File: src/tests/api/BavHappyPath.test.ts [hunk 4]
```diff
@@ -16,22 +17,27 @@ describe("Test BAV End Points", () => {
...
+        expect(authResponse.status).toBe(200);
+        expect(authResponse.data.authorizationCode).toBeTruthy();
+        expect(authResponse.data.redirect_uri).toBeTruthy();
+        expect(authResponse.data.state).toBeTruthy();
+
+        // Make sure authSession state is as expected
+        await getSessionAndVerifyKey(sessionId, constants.DEV_BAV_SESSION_TABLE_NAME, "authSessionState", "BAV_AUTH_CODE_ISSUED");
 
         // Make sure txma event is present & valid
         const sqsMessage = await getSqsEventList("txma/", sessionId, 1);
```
Comment (Z-Mohamoud) [inline src/tests/api/BavHappyPath.test.ts:37]: Why are we still only validating 1 TxMA message? 

The logic in place prior to this PR was hitting the session endpoint and validating **1** TxMA event. 

If you're adding an additional endpoint which results in an additional TxMA event we should be validating that we get 2 in total and both are formatted correctly.

File: src/tests/api/BavHappyPath.test.ts (hunk 3 above)
Comment (Z-Mohamoud) [inline src/tests/api/BavHappyPath.test.ts:22]: Please add your logic to a new test. 

I understand we can't test /authorization endpoint in completion due to the /verifyAccount endpoint not being in place. 

But we should not be adding this logic to the only BAV test we have working and marking it as skipped

File: src/tests/api/BavUnhappyPath.test.ts [hunk 5]
```diff
@@ -10,17 +12,32 @@ describe("/session Unhappy Path", () => {
         stubResponse = await stubStartPost(bavStubPayload);
     });
 
-    it("E2E Unhappy Path Journey - Invalid Request", async () => {
+    it.skip("E2E Unhappy Path Journey - Invalid Request", async () => {
```
Comment (Z-Mohamoud) [inline src/tests/api/BavUnhappyPath.test.ts:15]: Why is this being skipped?

File: src/tests/api/BavUnhappyPath.test.ts [hunk 6]
```diff
@@ -10,17 +12,32 @@ describe("/session Unhappy Path", () => {
...
         const sessionResponse = await sessionPost(stubResponse.data.clientId, "");
 
         expect(sessionResponse.status).toBe(401);
         expect(sessionResponse.data).toBe("Unauthorized");
     });
 
-    it("E2E Unhappy Path Journey - Invalid ClientID", async () => {
+    it.skip("E2E Unhappy Path Journey - Invalid ClientID", async () => {
```
Comment (Z-Mohamoud) [inline src/tests/api/BavUnhappyPath.test.ts:22]: Again, why is this being skipped?

File: src/tests/api/BavUnhappyPath.test.ts (hunk 2 above)
Comment (Z-Mohamoud) [inline src/tests/api/BavUnhappyPath.test.ts:38]: Another skipped test, should this not be working with what we currently have deployed?

File: src/tests/data/BAV_CRI_AUTH_CODE_ISSUED_SCHEMA.json [hunk 7]
```diff
@@ -0,0 +1,47 @@
+{
```
Comment (Z-Mohamoud) [inline src/tests/data/BAV_CRI_AUTH_CODE_ISSUED_SCHEMA.json:1]: Schema looks good, but if we're only validating 1 of the 2 TxMA events returned (BAV_CRI_START and BAV_CRI_AUTH_CODE_ISSUED) are you sure this schema is being used?

**Review by Z-Mohamoud — CHANGES_REQUESTED** [review]

File: src/tests/api/BavHappyPath.test.ts [hunk 8]
```diff
@@ -41,6 +73,26 @@ describe("Test BAV End Points", () => {
...
+
+describe("Happy Path Test for Repeat Authorisation", () => {
+    let sessionId: any;
+
+    beforeEach(async () => {
+        //Session Request
+        sessionId = await startStubServiceAndReturnSessionId(bavStubPayload);   
+    });
+
+    it.skip("Happy Path /authorisation endpoint - Repeated request made", async () => {
```
Comment (Caitlin-cooling) [inline src/tests/api/BavHappyPath.test.ts:85]: How come this is still skipped?

File: src/tests/api/BavHappyPath.test.ts [hunk 9]
```diff
@@ -16,22 +17,53 @@ describe("Test BAV End Points", () => {
...
+        // const userInfoResponse = await userInfoPost("Bearer " + tokenResponse.data.access_token);
+        // expect(userInfoResponse).toBe(202);
+    });
+});
 
-        // Commented out until /authorisation (KIWI-1259), /token (KIWI-1260) and /userInfo (KIWI-1258) endpoints are available
-        // // Authorization
-        // const authResponse = await authorizationGet(sessionId);
-        // expect(authResponse.status).toBe(200);
+describe("Test BAV End Points", () => {
```
Comment (Z-Mohamoud) [inline src/tests/api/BavHappyPath.test.ts:42]: Can we have all these tests under 1 description please, sharing the same beforeEach step as this is a pre-requisite for all happy path tests and differentiate in the it statement please? 

e.g. 

- Both tests can be under the same description

`describe("BAV CRI Happy Path", () => {`

- The beforeEach logic can stay the same 
- Differentiate between tests in the it step 
/session: `    it("BAV CRI /session Happy Path", async () => {`
/authorization: `    it("BAV CRI /authorization Happy Path", async () => {`

**Review by Z-Mohamoud — CHANGES_REQUESTED** [review]

**Comment by shenry99** [comment]
Closing so that branch can be merged into Dev KIWI-1256 branch.

**Pull request closed without merging**